src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs

```diff
@@ -1468,4 +1486,32 @@ private ProjectInstance CreateTestProject()
             return project.CreateProjectInstance();
         }
     }
+
+    /// <summary>
+    /// Task that returns a custom ITaskItem implementation that has a custom IDictionary type returned from CloneCustomMetadata()
+    /// </summary>
+    public sealed class TaskThatReturnsDictionaryTaskItem : Utilities.Task
+    {
+        public override bool Execute() => true;
+
+        [Output]
+        public ITaskItem DictionaryTaskItemOutput { get => new DictionaryTaskItem(); }
+    }
+
+    internal sealed class DictionaryTaskItem : ITaskItem
+    {
+        public string ItemSpec { get => $"{nameof(DictionaryTaskItem)}spec"; set => throw new NotImplementedException(); }
+
+        public ICollection MetadataNames => throw new NotImplementedException();
+
+        public int MetadataCount => throw new NotImplementedException();
+
+        private Dictionary<string, string> metaData = new() { ["a"] = "b" };
+
+        public IDictionary CloneCustomMetadata() => new Dictionary<string, string>(metaData);
```

Comment:
I believe we need a custom `IDictionary` implementation (not `Dictionary<string, string>`) here to test the https://github.com/dotnet/msbuild/pull/8870 fix